File: Java/src/solver/RungeKuttaSolver.java
// RungeKuttaSolver.java
// Class implementing the 4th order Runge-Kutta method for solving ODEs

package solver;

import java.util.List;
import java.util.ArrayList;

public class RungeKuttaSolver {
    private DifferentialEquation equation;
    private double initialY;

    // Constructor
    public RungeKuttaSolver(DifferentialEquation equation, double initialY) {
        this.equation = equation;
        this.initialY = initialY;
    }

    // Solves the differential equation for given time points
    public List<Double> solve(List<Double> t) {
        List<Double> y = new ArrayList<>();
        y.add(initialY);

        for (int i = 1; i < t.size(); i++) {
            double h = t.get(i) - t.get(i - 1);
            double yPrev = y.get(i - 1);
            double tPrev = t.get(i - 1);

            double k1 = h * equation.evaluate(tPrev, yPrev);
            double k2 = h * equation.evaluate(tPrev + h / 2, yPrev + k1 / 2);
            double k3 = h * equation.evaluate(tPrev + h / 2, yPrev + k2 / 2);
            double k4 = h * equation.evaluate(tPrev + h, yPrev + k3);

            double yNext = yPrev + (k1 + 2 * k2 + 2 * k3 + k4) / 6;
            y.add(yNext);
        }

        return y;
    }
}

File: C++/src/RungeKuttaSolver.cpp
// RungeKuttaSolver.cpp
// Implementation of the RungeKuttaSolver class

#include "RungeKuttaSolver.h"

RungeKuttaSolver::RungeKuttaSolver(Func f, double initialY)
    : f(f), initialY(initialY) {}

std::vector<double> RungeKuttaSolver::solve(const std::vector<double>& t) {
    int n = t.size();
    std::vector<double> y(n);
    y[0] = initialY;

    for (int i = 1; i < n; ++i) {
        double h = t[i] - t[i - 1];
        double k1 = h * f(t[i - 1], y[i - 1]);
        double k2 = h * f(t[i - 1] + h / 2, y[i - 1] + k1 / 2);
        double k3 = h * f(t[i - 1] + h / 2, y[i - 1] + k2 / 2);
        double k4 = h * f(t[i - 1] + h, y[i - 1] + k3);

        y[i] = y[i - 1] + (k1 + 2 * k2 + 2 * k3 + k4) / 6;
    }

    return y;
}
